.github/workflows/validate-commit.yml

```diff
@@ -41,8 +41,8 @@ jobs:
         path: |
           ~/.cargo/registry
           ~/.cargo/git
-          target
-        key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}-${{ secrets.CI_CACHE_VERSION }}
+          target/debug
+        key: lint-${{ runner.os }}-${{ hashFiles('**/Cargo.lock') }}-${{ secrets.CI_CACHE_VERSION }}
```

Comment:
Note that the `lint` and `test` jobs now have separate caches.

The old setup had a problem. First, some context:
- `lint` only needs sources; after running it `target/` doesn't contain any compiled artifacts.
- `test` obviously does compile stuff, and is much faster when `target/` already contains compilation artifacts.
-  `lint` is much quicker than `test`. They run concurrently in this workflow with `lint` usually finishing first.
- `actions/cache@v2` does not update the cache if one already exists for the computed key, even if there are changes in the cached paths. Cache entries are immutable and don't accumulate changes as more jobs are run. (See https://github.com/actions/cache/issues/342 too.)

The problem was that with a cold cache, `lint` finishes before `test` and therefore is the one to create the cache entry, but at that point we only have the sources. When `test` finishes the compiled stuff gets thrown away, even on its first run. So it needs to re-compile every time. Funnily enough, the `macos-latest` test run was immune to this: `lint` is only run on `ubuntu-latest` so the mac tests weren't racing against anything to create that cache.

Separating the caches for `lint` and `test` sidesteps this problem.

Reply:
The lesson here is that you should think about the arrangement of the workflow DAG when coming up with cache keys. Know which job should be responsible for creating the cache entry and make sure it doesn't run concurrently with other jobs using that cache

Sharing a cache key between jobs (as in release.yml) is only okay if the jobs form a sub-DAG with a single root. The root job will be the one who creates the cache entry.